### PR TITLE
Make optimizer usable in Docker and improve its logging

### LINKED ISSUE #5256: improve local optimization
#### Expected Behavior

- The optimization launcher finishes when optimization is done
- An overview / mapping of the parameter set <-> backtest id should be provided for further processing (eg plotting)
- A config can be provided for optimization tasks

I have already implemented these improvements so that they work for me (see link below).
Please let me know how I should edit my changes, so that others can benefit.

#### Actual Behavior

- The optimization launcher immediately returns when run from docker

#### Potential Solution
[see this comparison](https://github.com/QuantConnect/Lean/compare/115dcab...radi-ka:feature/local_optimization_improvements)

#### Reproducing the Problem
```shell
docker run \
 --rm \
 --mount type=bind,source="$LOCAL_OPT_CONFIG",target=/Lean/Launcher/bin/Debug/config.json,readonly \
 --mount type=bind,source="$LOCAL_DATA",target=/Data \
 --mount type=bind,source="$LOCAL_RESULTS",target=/Results \
 --mount type=bind,source="$LOCAL_ALGO_DLL",target=/Lean/Launcher/bin/Debug/QuantConnect.Algorithm.CSharp.dll,readonly \
 --name LeanEngine-1 \
 --entrypoint mono \
 -p 55555:55555 \
 quantconnect/lean \
 QuantConnect.Optimizer.Launcher.exe --config /Lean/Launcher/bin/Debug/config.json
```

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by jmerle**
#### Description

- Updated the Dockerfile so that the optimizer's launcher is added to the quantconnect/lean:latest image.
- Made the optimizer's launcher wait until the optimizer has stopped running before exiting.
- Made the optimizer's launcher save its logging to `<results-destination-folder>/log.txt`.
- Changed the `launched backtest '<backtest id>'` messages to `launched backtest '<backtest id>' with parameters '<parameters>'` so that the generated log.txt file contains a backtest id <> parameter set mapping.

#### Related Issue

Closes #5256.

#### Motivation and Context

This PR makes the optimizer usable in Docker.

#### Requires Documentation Change

No.

#### How Has This Been Tested?

Tested with Mono locally as follows:
```
$ msbuild QuantConnect.Lean.sln
$ cd Optimizer.Launcher/bin/Debug
$ mono QuantConnect.Optimizer.Launcher.exe
```

Tested with Mono in Docker as follows:
```
$ msbuild QuantConnect.Lean.sln
$ docker build -t lean:local .
$ docker run --rm \
    --mount type=bind,source=/home/jasper/Projects/Lean/Launcher/config.json,target=/Lean/Launcher/bin/Debug/config.json,readonly \
    --mount type=bind,source=/home/jasper/Projects/Lean/Optimizer.Launcher/config.json,target=/Lean/Optimizer.Launcher/bin/Debug/config.json,readonly \
    -v /home/jasper/Projects/Lean/Data:/Lean/Data:ro \
    -w /Lean/Optimizer.Launcher/bin/Debug \
    --entrypoint mono \
    lean:local \
    QuantConnect.Optimizer.Launcher.exe
```

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Optimizer.Launcher/Program.cs [hunk 1]
```diff
@@ -63,7 +68,11 @@ public static void Main()
                 Log.Error(e);
             }
 
-            Console.ReadKey();
+            // Wait until the optimizer has stopped running before exiting
+            while (optimizer != null && (optimizer.Status == OptimizationStatus.New || optimizer.Status == OptimizationStatus.Running))
+            {
+                Thread.Sleep(100);
```
Comment (Martin-Molinero) [inline Optimizer.Launcher/Program.cs:74]: I think we could directly create a reset event and wait on it and set it in the `optimizer.Ended` callback after disposing of the `optimizer`? So we directly don't need to poll or expose the state

**Review by Martin-Molinero — CHANGES_REQUESTED** [review]
Nice! Minor request 👍

File: Optimizer.Launcher/Program.cs (hunk 1 above)
Comment (jmerle) [reply Optimizer.Launcher/Program.cs]: Didn't know C# had such a feature, fixed :+1:

**Review by Martin-Molinero — APPROVED** [review]
Thanks!

**Pull request merged**